### PR TITLE
feat: Add support for the new `ChatMessage` data class in `ChatPromptBuilder`

### PR DESCRIPTION
**Pull request by shadeMe**
### Proposed Changes:

 
 
Since the newly-introduced `ChatMessage` dataclass in experimental defers in type signature, components in the core package that consume the above dataclass are no longer able to use the new version due to the type-checking constraints of the pipeline and breaking changes to the class.

This PR updates one such component to use the new version of the `ChatMessage` dataclass. Other components will be duly updated once tool support has been merged into the core package.

### How did you test it?

Unit tests

### Checklist

- I have read the [contributors guidelines](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack/blob/main/code_of_conduct.txt)
- I have updated the related issue with new insights and changes
- I added unit tests and updated the docstrings
- I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.
- I documented my code
- I ran [pre-commit hooks](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md#installation) and fixed any issue

### CODE REVIEW

File: haystack_experimental/components/builders/chat_prompt_builder.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+    def _validate_variables(self, provided_variables: Set[str]):
+        """
+        Checks if all the required template variables are provided.
+
+        :param provided_variables:
+            A set of provided template variables.
+        :raises ValueError:
+            If no template is provided or if all the required template variables are not provided.
+        """
+        missing_variables = [var for var in self.required_variables if var not in provided_variables]
```
Comment (anakin87) [inline haystack_experimental/components/builders/chat_prompt_builder.py:227]: do we want to include the change introduced in https://github.com/deepset-ai/haystack/pull/8572?

File: haystack_experimental/dataclasses/chat_message.py [hunk 2]
```diff
@@ -158,24 +158,34 @@ def is_from(self, role: ChatRole) -> bool:
...
+    def from_user(
+        cls,
+        text: str,
+        meta: Optional[Dict[str, Any]] = None,
+    ) -> "ChatMessage":
         """
         Create a message from the user.
 
         :param text: The text content of the message.
+        :param meta: Additional metadata associated with the message.
```
Comment (anakin87) [inline haystack_experimental/dataclasses/chat_message.py:170]: In the current implementation of `ChatMessage` in Haystack, metadata are only supported in messages with assistant role.
So I replicated this in the experimental dataclass.
Do you think it is more consistent to allow metadata in `ChatMessage` for each possible role?

**Review by anakin87 — COMMENTED** [review]
Looks good in general. I left two minor comments.

I see that "Sync docs with Readme" workflow is failing, but I don't know if it's related to the current PR.

**Comment by shadeMe** [comment]
> I see that "Sync docs with Readme" workflow is failing, but I don't know if it's related to the current PR.

That's unrelated; has to do with the fact that this PR comes from a fork.

**Review by anakin87 — APPROVED** [review]

**Pull request merged**